core/src/Serial/Kokkos_Serial_Parallel_Range.hpp

```diff
@@ -191,10 +200,16 @@ class ParallelScan<FunctorType, Kokkos::RangePolicy<Traits...>,
     auto* internal_instance = m_policy.space().impl_internal_space_instance();
     std::lock_guard<std::mutex> lock(
         internal_instance->m_thread_team_data_mutex);
+
     internal_instance->resize_thread_team_data(
         pool_reduce_size, team_reduce_size, team_shared_size,
         thread_local_size);
 
+    // Make sure kernels are running sequentially even when using multiple
+    // threads
+    std::lock_guard<std::mutex> instance_lock(
+        internal_instance->m_instance_mutex);
```

Comment:
You probably should lock 1st thing or comment why you deferred until after the resize.
Same comment below.

Reply:
We need both lock the `m_thread_team_data_mutex` and the `m_instance_mutex` since we want to make sure that there is no other kernel that modifies the scratch allocations while this kernel is executing.
We can't just use the `m_instance_mutex` alone (before `resize_thread_team_data`) since that function might call `deallocate` which implies a global fence and that fence tries locking `instance_lock` resulting in a deadlock.

Reply:
```
    frame #12: 0x000000010066c384 Kokkos_CoreUnitTest_Serial1`Kokkos::fence(name="HostSpace::impl_deallocate before free") at Kokkos_Core.cpp:1113:47 [opt]
    frame #13: 0x000000010066d60c Kokkos_CoreUnitTest_Serial1`Kokkos::HostSpace::impl_deallocate(this=<unavailable>, arg_label="Kokkos::Serial::scratch_mem", arg_alloc_ptr=0x000000013980b800, arg_alloc_size=11264, arg_logical_size=0, arg_handle=(name = "Host")) const at Kokkos_HostSpace.cpp:122:5 [opt]
    frame #14: 0x000000010066d560 Kokkos_CoreUnitTest_Serial1`Kokkos::HostSpace::deallocate(this=<unavailable>, arg_label=<unavailable>, arg_alloc_ptr=<unavailable>, arg_alloc_size=<unavailable>, arg_logical_size=<unavailable>) const at Kokkos_HostSpace.cpp:115:3 [opt]
    frame #15: 0x00000001006773bc Kokkos_CoreUnitTest_Serial1`Kokkos::Impl::SerialInternal::resize_thread_team_data(this=0x0000600001798000, pool_reduce_bytes=512, team_reduce_bytes=512, team_shared_bytes=8, thread_local_bytes=0) at Kokkos_Serial.cpp:119:13 [opt]
    frame #16: 0x00000001000477e0 Kokkos_CoreUnitTest_Serial1`Kokkos::Impl::ParallelFor<Test::TestSharedAtomicsFunctor<Kokkos::Serial>, Kokkos::TeamPolicy<Kokkos::Serial>, Kokkos::Serial>::execute(this=0x000000016fdfefe0) const at Kokkos_Serial_Parallel_Team.hpp:255:24 [opt]
    frame #17: 0x0000000100047634 Kokkos_CoreUnitTest_Serial1`void Kokkos::parallel_for<Kokkos::TeamPolicy<Kokkos::Serial>, Test::TestSharedAtomicsFunctor<Kokkos::Serial>, void>(str=<unavailable>, policy=0x000000016fdff0d0, functor=0x000000016fdff080) at Kokkos_Parallel.hpp:144:11 [opt]
    frame #18: 0x0000000100045e30 Kokkos_CoreUnitTest_Serial1`Test::serial_atomic_shared_Test::TestBody() [inlined] void Kokkos::parallel_for<Kokkos::TeamPolicy<Kokkos::Serial>, Test::TestSharedAtomicsFunctor<Kokkos::Serial>>(policy=0x000000016fdff0d0, functor=0x000000016fdff080, (null)=0x0000000000000000) at Kokkos_Parallel.hpp:153:3 [opt]
    frame #19: 0x0000000100045e18 Kokkos_CoreUnitTest_Serial1`Test::serial_atomic_shared_Test::TestBody(this=<unavailable>) at TestAtomicOperations_shared.hpp:49:3 [opt]
    frame #20: 0x000000010063cb14 Kokkos_CoreUnitTest_Serial1`void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) [inlined] void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(object=<unavailable>, method=0x00000000000000010000000000000020, location="the test body") at gtest-all.cc:4101:10 [opt]
    frame #21: 0x000000010063cb04 Kokkos_CoreUnitTest_Serial1`void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(object=0x0000600002e88100, method=<unavailable>, location="the test body") at gtest-all.cc:4137:14 [opt]
    frame #22: 0x000000010063ca68 Kokkos_CoreUnitTest_Serial1`testing::Test::Run(this=0x0000600002e88100) at gtest-all.cc:4176:5 [opt]
    frame #23: 0x000000010063e1d4 Kokkos_CoreUnitTest_Serial1`testing::TestInfo::Run(this=0x0000000139605a00) at gtest-all.cc:4326:11 [opt]
    frame #24: 0x000000010063ed90 Kokkos_CoreUnitTest_Serial1`testing::TestSuite::Run(this=0x00000001396048f0) at gtest-all.cc:4480:28 [opt]
    frame #25: 0x000000010064eb74 Kokkos_CoreUnitTest_Serial1`testing::internal::UnitTestImpl::RunAllTests(this=0x000000013a804080) at gtest-all.cc:7320:44 [opt]
    frame #26: 0x000000010064e208 Kokkos_CoreUnitTest_Serial1`bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) [inlined] bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(object=<unavailable>, method=(Kokkos_CoreUnitTest_Serial1`testing::internal::UnitTestImpl::RunAllTests() at gtest-all.cc:7211), location="auxiliary test code (environments or event listeners)") at gtest-all.cc:4101:10 [opt]
    frame #27: 0x000000010064e1f8 Kokkos_CoreUnitTest_Serial1`bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(object=0x000000013a804080, method=<unavailable>, location="auxiliary test code (environments or event listeners)") at gtest-all.cc:4137:14 [opt]
    frame #28: 0x000000010064e178 Kokkos_CoreUnitTest_Serial1`testing::UnitTest::Run(this=0x0000000100854948) at gtest-all.cc:6903:10 [opt]
    frame #29: 0x000000010000369c Kokkos_CoreUnitTest_Serial1`main [inlined] RUN_ALL_TESTS() at gtest.h:12372:46 [opt]
    frame #30: 0x0000000100003694 Kokkos_CoreUnitTest_Serial1`main(argc=1, argv=0x000000016fdff638) at UnitTestMainInit.cpp:26:16 [opt]
    frame #31: 0x0000000189ebe0e0 dyld`start + 2360
```

Reply:
@Rombur The above backtrace shows that we are calling `Kokkos::fence` in `deallocate` and that calls the static fence that is introduced in this pull request.

Reply:
I saw that but my understanding is that the global fence was added to fix a deadlock in the `Threads` backend and the code has changed quite a bit since then. I want to understand if the fence is still necessary or if we could do something in the `Threads` backend, so that we don't pollute the other backend.

Reply:
We need the fence (introduced in https://github.com/kokkos/kokkos/pull/4499) for asynchronous host backends like `HPX` so we don't free an allocation that is still used (same reasoning as for the GPU backends). As for the scope of then fence, see https://github.com/kokkos/kokkos/pull/4499/commits/69ce416956a2350de03f8f81206e5f042f834bcc#r747018936.

Reply:
I'm open to rediscussing that `deallocate` only fences asynchronous execution spaces (that can access the respective allocation).